### ISSUE TITLE
inlining types

### ISSUE THREAD

**Issue by KOLANICH** [kaitai-io/kaitai_struct#88]
```yaml
types:
 str_ptr:
  seq:
  - id: ptr
     type: u2
  instances:
    str:
      pos: ptr
      type: strz
```

Instantiation of this can be transformed into nasty code. So we need some attributes to show that in this case we don't need an object of type str_ptr but need only string

**Comment by generalmimon** [comment]
This would be a great feature and seems quite easy to implement, although there are some questions unanswered about this issue.

## Syntax

I vote for something like `passthrough-value` suggested by @GreyCat in https://github.com/kaitai-io/kaitai_struct/issues/119#issuecomment-283000464 (but I don't like the name, hope that someone will come with something shorter and more comprehensible):

```yaml
seq:
  - id: subrecords_array
    type: subrecord_array
types:
  subrecord_array:
    seq:
      - id: subrecords
        type: subrecord
        repeat: eos
    passthrough-value: subrecords
```

Compiler would generate this automatically:
```java
    public ArrayList<Subrecord> subrecordsArray() { return subrecordsArray.subrecords(); }
```

Why I prefer a type-level key referencing the field to inline over adding `inline: true` to the field level?

  1. It makes sense to inline only one field in the type. With `inline: true`, the compiler would need to **iterate** all fields (`seq` attributes, value and parse instances) and check if at most once is present, otherwise throw an error. That's definitely harder than checking a single key like `passthrough-value`.

  2. (Related to 1.) It loses declarativeness, because to check that the KSY is well-formed, you have to make a loop with some **collector** variable (which seems imperative). I don't even know if that is possible to enforce with the JSON schema (to be able to check the correctness with https://github.com/kaitai-io/ksy_schema/), but I don't think so.

  3. It's easier to read. Instead of finding "where the hell is the `inline: true`" when reviewing some type with lots of `seq` and `instance` fields, you just see `passthrough-value` (we can recommend in style guide to put it up or down in the type definition, not sure what'll to be better) and understand what's going on.

  4. Inlining doesn't change the way how the field is parsed, it influences how the **whole type** behaves, hence it should be present on the type level.


However, with `passthrough-value` syntax comes one question which doesn't apply to attr-level `inline` key: should it allow some complex expressions like `passthrough-value: _parent.field * 2` instead of just referencing an attribute that is present in the current struct?

I definitely vote for disallowing to use any expressions in the `passthrough-value` (hmm, then `passthrough-value` is misleading, it should be called `passthrough-field` or `passthrough-attr`, but still terrible names :smiley:) except for plain identifier of a field in the same struct. If you needed to inline a processed value with some expression, it should be done by creating a value instance and referencing it. The reason is serialization. Take for example this spec:

```yaml
seq:
  - id: usage
    type: triple
types:
  triple:
    seq:
      - id: int
        type: u1
        dependent: val / 3
    instances:
      val:
        value: int * 3
    passthrough-attr: val
```

I use [`dependent` key here from serialization proposal](https://github.com/kaitai-io/kaitai_struct/issues/27#issuecomment-289577179) to mark the `int` as read-only field (`setInt` won't be generated) and introduce a formula that calculates value of this field from `val` (i.e. from the inlined value of the `triple` type). This is very nice because you can just call `setUsage(30)` and you don't have to bother with setting the `int` field programmatically, it gets set to `10` automatically. If we would allow doing `passthrough-value: int * 3` instead, this wouldn't work because we couldn't reference the value of `passthrough-value` from the `dependent` clause.

## Serialization

Actually, the most serious thing I'm concerned about this feature is serialization. Imagine the type above brought by @KOLANICH:

```yaml
seq:
  - id: aaaa
    type: str_ptr
types:
  str_ptr:
    seq:
      - id: ptr
        type: u2
    instances:
      str:
        pos: ptr
        type: strz
    passthrough-attr: str
```

Compiler would generate these accessors automatically:
```java
    public String aaaa() { return aaaa.str(); }
    public void setAaaa(String _v) { aaaa.setStr(_v); }
```

As an author of the KSY, you're probably OK with permanently losing access to `ptr` after reading the `str`. However, as a programmer which needs to serialize this structure and to move this attribute to another offset in the stream, you're not. You need to have a public setter to change the `ptr` attribute to be able to do that.

There are two use cases that are important: mutating the state of the current `str_ptr` struct and creating a new `str_ptr` object from scratch. They need both getter and setter for the `str_ptr` struct, respectively. It could look like this:

```java
    public StrPtr _aaaa() { return aaaa; }
    public void _setAaaa(StrPtr _v) { aaaa = _v; }
```

Prefixing these methods with an underscore is ugly, but I don't know how else to avoid the name collision.

**Comment by KOLANICH** [comment]
Thanks for the verbose response. I aggree to almost every point.

1. `passthrough-value` seems to be too long. I guess `inline: val` is no worse.
2. We don't have to store `ptr` after we have discarded it. If we discard something we `don't care` about it, so the impl is free to recreate that from scratch when serializing as it wishes, if the result of parsing of the serialization will be the same. Some more constraints can be added via asserts.

**Comment by Mingun** [comment]
`passthrough-value` can be just named `value`, because its intention is to return some part of type instead of all its structure, i.e. redefine final value

> They need both getter and setter for the str_ptr struct, respectively. It could look like this:

Yes, just add some prefix to the field. This is commonly used pattern, see for example, [XmlBeans](https://xmlbeans.apache.org/docs/3.0.0/guide/conMethodsForGeneratedJavaTypes.html) -- there is `[get|set]Name` and `[xget|xset]Name`

**Comment by generalmimon** [comment]
@Mingun:
> `passthrough-value` can be just named `value`, because its intention is to return some part of type instead of all its structure, i.e. redefine final value

I disagree with this. `value` is already used for specifying the expression to calculate the value of a value instance. The `passthrough-field` (or the type-level `inline` suggested by @KOLANICH, I'm fine with that) is somewhat different. Its value is not an KS expression, but only a name of a attribute in the same struct (not even `_parent._parent.field`, just `some_field`, the reason I discuss in the previous comment). Using `value` for inlining suggests that expressions can be used as in case of calculated instances.

The second reason is that if we name it different from `value`, you can just use fulltext search for `inline: ` and it highlights you only the inlined field of the type.

In general, I strongly disagree naming different things the same and vice versa. If the contents of `value` would be expression on instance level and plain name of field on type level, they're clearly not the same things.

**Comment by Mingun** [comment]
> I disagree with this. `value` is already used for specifying the expression to calculate the value of a value instance.

Exactly. This is main motivation to use it. There is no difference in semantics.

> If we would allow doing `passthrough-value: int * 3` instead, this wouldn't work because we couldn't reference the value of `passthrough-value` from the dependent clause.

Of course can:
```yaml
seq:
  - id: usage
    type: triple
types:
  triple:
    seq:
      - id: int
        type: u1
        dependent: _ / 3 # but I think it is more logical name to attribute as `write`
    value: int * 3
```

**Comment by KOLANICH** [comment]
BTW, there is an important issue here: inlined types must be inlineable in compile time. Otherwise KS runtime will have to be a virtual machine - we can implement a recursive inline reduced to a single variable, this way implementing a recursive function. Making KS runtimes Turing-complete opens cans of worms, affecting both security and serializability. So, no loops in inlineable types graph should be possible.

**Comment by dgelessus** [comment]
@generalmimon Thank you for the detailed write-up!

Regarding the name, I don't have a strong preference, but I would rather have something more descriptive (I would even be okay with a long name like `passthrough-attr`). You only have to type the attribute once per type (and I'd guess that most types won't use this feature), so in this case readability is more important than writing speed IMHO.

I agree that there should be some way to "bypass" the inlining, so that the underlying type's real members can still be accessed. This should be possible both in KSY and in application code that uses the compiled code. @generalmimon's comment describes how this is needed for serialization, but there are other use cases too: visualizers should be able to display the real structure behind inlined types, and if the inlined member is a value instance you might want to bypass it (if it's slow or can fail for some inputs).

Not sure what the naming convention could be for "bypassing" inlining. Just an underscore won't be enough (otherwise you get conflicts when a member called `io` is inlined). Something like `_inlined_<originalname>` could work. In terms of discoverability and usability, something like `<originalname>_inlined` would be better (so it shows up in code completion next to the inlined member), but then you have potential naming conflicts again.

**Comment by KOLANICH** [comment]
>Not sure what the naming convention could be for "bypassing" inlining.

I propose to just have a special compile flag for that.
`--noinline=".*"` to disable inlining globally
`--noinline=yamlTypePathRx,yamlPropertyPathRx` to disable only for certain types and properties.

**Comment by dgelessus** [comment]
> I propose to just have a special compile flag for that.
> `--noinline=".*"` to disable inlining globally
> `--noinline=yamlTypePathRx,yamlPropertyPathRx` to disable only for certain types and properties.

Hm, I'm not a big fan of that. This might work fine for visualizers, but not for other use cases where most of the time you want to take advantage of the inlining, and only sometimes you want to bypass it and access the underlying members.

Also, I'd expect visualizers to display inlined members differently, for example the WebIDE would treat `inline`/`passthrough-attr` as an implicit `-webide-representation`. (This would also somewhat address the readability issues brought up by @Mingun in #666.) If you tell the compiler to skip inlining completely, that won't be possible anymore, because the inlined-ness of types won't be visible to the visualizer.

**Comment by KOLANICH** [comment]
>but not for other use cases where most of the time you want to take advantage of the inlining, and only sometimes you want to bypass it and access the underlying members.

You are right, we shouldn't disable inlining, it would break the specs that use the inlined types. We definitely should save additional info. For python and js we already do it, there already exist a special property for that.

But the most important point is that we shouldn't do it by default, and when enabled, we shouldn't do it for all the inlined types. I mean that one of the points of inlining is reducing memory usage. Let's assumme someone has written a spec.

For example there is an extremily widespread security vulnr when software developers forget to wipe the memory before putting a string into a fixed-size field. #613   It is possible to write a type in KS accessing both the content and the garbage.  Normally when one parses such a format he doesn't need this garbage. But for some apps one need to access this garbage. But if one wants to access the garbage it doesn't mean he needs access to all the other discarded stuff. So wheter the discarded stuff is needed and which one is needed should be decided by the one using the spec.

**Comment by dgelessus** [comment]
> I mean that one of the points of inlining is reducing memory usage.

I wasn't aware of that - from the previous comments I had the impression that inlining (as proposed in this issue) is just syntactical sugar for better usability.

Discarding the values of some fields after parsing to optimize memory usage may be useful, but I don't think it should be coupled with inlining - I would implement it as a separate attribute. Depending on the format, you may want to use inlining to get a nicer API, but still have the ability to access an inlined type's internal fields from inside the KSY.

**Comment by KOLANICH** [comment]
How about the following table?

| inline is set | type-level discard | prop-level-discard | effect on a discardable prop |
| - | - | - | - |
| 0 | 0 | 0 | as usual |
| 1 | z | z | discard |
| X | X | 1 | discard |
| X | X | 0 | don't discard |
| X | 1 | z | discard |
| X | 0 | z | don't discard |

`X` - any, [`z`](https://en.wikipedia.org/wiki/Three-state_logic) - the prop is missing

**Comment by dgelessus** [comment]
> How about the following table?

I'm not sure what you're asking here, sorry. The table on its own doesn't make much sense to me - what are `discard` and `no-discard` and what is a "discardable prop"? If this is a proposal for how property discarding should be controlled, I think it would be better to make a separate issue for that with some more context.

**Comment by KOLANICH** [comment]
>If this is a proposal for how property discarding should be controlled, I think it would be better to make a separate issue for that with some more context.

I propose to closely couple discarding with inlining. This table can be summarized as:
* when inlined, by default discard
* it is forbidden to mark unused props as not discarded if inlining is enabled
* prop-level discarding has priority over type-level discarding
* never discard selected via a CLI switch

**Comment by goyalyashpal** [comment]
i can't properly understand the issue-description.  what is current scenario, what is proposal?
sectioning these under different headings helps a lot.

the common headings i use are these, but feel free to create your own:  
`summary`, `current scenario`, `problem` (affects of current scenario), `solution proposal`

**Comment by goyalyashpal** [comment]
_self hiding as offtopic_

----
> For example there is an extremily widespread security vulnr when software developers forget to wipe the memory before putting a string into a fixed-size field.
>
> \- _@ KOLANICH at https://github.com/kaitai-io/kaitai_struct/issues/88#issuecomment-575899065_

wow, true.  i just faced this bug / vulnerability in an application.  a long string was put in a fixed sized field & its remaining content was partially exposed in other string fields.

**Comment by goyalyashpal** [comment]
### summary

allow some way to specify type-aliases for elemental types


### details

[subtypes]: https://doc.kaitai.io/user_guide.html#types "4.8. Substructures (subtypes)"
[instances]: https://doc.kaitai.io/user_guide.html#_instances_data_beyond_the_sequence "4.13. Instances: data beyond the sequence"

**current situation:**
* kaitai allows specifying [subtypes], but not "type aliases"
* with type-alias, i mean the parsing specification (e.g. `type`, `size`, `enum`, `valid`, `encoding`, etc.) for _single field value_ as against the subtype (defined in `types.<id>.seq`) which is for multiple fields


**use scenario:**

[dry]: https://en.wikipedia.org/wiki/Dont_repeat_yourself

* on using `types` construct to define single-field type-aliases, 
* ... it gets needlessly parsed as a subtype (thus, redundant sub-attribute / children node etc) with name `_unnamedX`
* this looks ugly, increases output's size, & require meaningless unfolding / exapnding
*  <details><summary>screenshots:</summary>
    <br/>
    <img width="300" alt="Image" src="https://github.com/user-attachments/assets/16fb6f44-3cc2-4a29-b51d-9ec08ba7e0ca" /> <img width="300"  alt="Image" src="https://github.com/user-attachments/assets/19d97ad1-1247-4e60-aaf3-94b65487b6a1" />

  </details>




`.ksy` samples:

* small example:  `types: {my_type_alias: {seq: [{type: u1, enum: my_enum}]}}`

* <details><summary>working sample with multiple aliases:</summary>
  <p>

  ```yaml
  types:

    my_type_alias_a:
      seq:
        - {type: u1, enum: my_enum}
    
    my_file_name:  # alias for elemental type "string"
      seq:
        - {type: str, size: 10, encoding: utf-8}  # why does ksc complain about utf-8's lower-case?
    
    my_validation:
      seq:
        - {type: u4le, valid: {min: 10}}

  ```

  </p>
  </details>

**additional info:**
type aliases are used in a _lot_ of places:
* as above schema example shows: for specifying `enum`, `size`, `valid`, or any myriad of attributes
* for reusing same specification at multiple places i.e. [DRY]

for these uses, there really is no need of any `seq`uence, as all the definition is for single value only.

also, these two features (type-alias, and sequence-def) are fairly orthogonal i think, and can be easily composed on user's end if the need arises.


### solutions

<details><summary>original</summary>

<strike>
<sub>
* i am not sure at the moment regarding what the details of implementation (even for syntax) can be, considering the cross-language exports.
<br/>* i thought about [instances], but it seems it requires explicitly specifying `pos`, and it's purpose seems something else too.
</sub></strike>

</details>

### versions

```
Kaitai WebIDE version: 0.1-SNAPSHOT-a02d396 (2025-09-06 13:13:49, license)
Kaitai compiler version: 0.11.0 (2025-09-07T18:59:46Z)
```


### meta

[type alias]: https://github.com/kaitai-io/kaitai_struct/issues?q=is%3Aissue%20type%20alias
[typedef]: https://github.com/kaitai-io/kaitai_struct/issues?q=is%3Aissue%20typedef
[element type]: https://github.com/kaitai-io/kaitai_struct/issues?q=is%3Aissue%20element%20type


i tried searching for existing duplicates with following queries:
* [type alias]
* [typedef]
* [element type]



----
[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: ../issues?q=is%3Aissue+state%3Aopen+sort%3Areactions-%2B1-desc

Add a 👍 [reaction] to [issues you find important].

**Comment by EmperorArthur** [comment]
+1 This provides built in support for an extremely common pattern I use to work around other limitations:

Simple example:
```yaml
split_data:
  seq:
  - id: raw_repeated_1
    type: u1
    repeat: until
    repeat-until: _ == 0xff
  - id: raw_repeated_2
    type: u1
    repeat: until
    repeat-until: _ == 0xff
  instances:
    sums:
      doc: `repeat` cannot be used with `value`.
      type: indexed_value(raw_repeated_1[_index]+raw_repeated_2[_index])
      repeat: expr
      repeat-expr: raw_repeated_1.size - 1
    repeated_1:
      doc:`include: false` does not work with `repeat`
      type: indexed_value(repeated_1)
      repeat: expr
      repeat-expr: raw_repeated_1.size - 1

indexed_value:
  params:
  - id: value
    type: any
```

```python
def unwrap(obj):
  if isinstance(obj, Iterable):
    return list(map(unwrap, obj))
  if hasattr(obj, 'value'):
    return obj.value
  return obj
```
